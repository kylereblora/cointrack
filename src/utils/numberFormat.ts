const defaultNumberFormatConfig = {
  minimumFractionDigits: 2,
};

const currencyFormatter = (currency = "PHP") =>
  new Intl.NumberFormat("en-US", {
    style: "currency",
    currency,
    ...defaultNumberFormatConfig,
  });

const formatPercentage = (value: string | number) => {
  const formatter = new Intl.NumberFormat("en-US", {
    ...defaultNumberFormatConfig,
  });

  if (typeof value === "string") {
    return `${formatter.format(parseFloat(value) * 100)}%`;
  }

  return `${formatter.format(value * 100)}%`;
};

const formatCurrency = (value: string | number, currency = "PHP") => {
  if (typeof value === "string") {
    return currencyFormatter(currency).format(parseFloat(value));
  }

  return currencyFormatter(currency).format(value);
};
export { formatPercentage, formatCurrency };
