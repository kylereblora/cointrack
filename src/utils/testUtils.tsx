import * as React from "react";
import { render as rtlRender, RenderOptions } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CurrencyProvider } from "../context/currencyContext";
import { ThemeProvider } from "../context/themeContext";
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";
import { SWRConfig } from "swr";

type WrapperProps = {
  children: React.ReactNode;
};

// solves the issue raised in https://github.com/emotion-js/emotion/issues/1105
const emotionCache = createCache({ key: "emotion-test-util-key" });
emotionCache.compat = true;

function Wrapper({ children }: WrapperProps): React.ReactElement {
  // since we're using useSWR, it's caching does not play out well when testing the app in isolation
  // and might bleed out to other tests. To fix this, we'll render SWRConfig with dedupingInterval: 0
  // in the wrapper and add cache.clear() to setupTest.ts
  return (
    <SWRConfig value={{ dedupingInterval: 0 }}>
      <CacheProvider value={emotionCache}>
        <CurrencyProvider>
          <ThemeProvider>{children}</ThemeProvider>
        </CurrencyProvider>
      </CacheProvider>
    </SWRConfig>
  );
}

// custom render util with a custom wrapper
function render(ui: React.ReactElement, options?: RenderOptions) {
  return rtlRender(ui, { wrapper: Wrapper as React.ComponentType, ...options });
}

export * from "@testing-library/react";

export { render, userEvent };
