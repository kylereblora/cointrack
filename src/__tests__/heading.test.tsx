import * as React from "react";
import { render, screen, userEvent } from "../utils/testUtils";
import Header from "../components/header";

test("renders the header component properly", () => {
  const { container } = render(<Header />);
  expect(container).toMatchSnapshot();
});

test("should update fiat conversion when dropdown is pressed", () => {
  render(<Header />);

  // get the dropdown button element, then press it
  const dropdown = screen.getByRole("button");
  userEvent.click(dropdown);

  // get an option inside the dropdown list, doesn't matter which
  const fiat = screen.getByRole("option", { name: /EUR/ });
  userEvent.click(fiat);

  expect(dropdown).toHaveTextContent(/EUR/);
});
