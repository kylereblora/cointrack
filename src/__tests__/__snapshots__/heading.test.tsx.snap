// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the header component properly 1`] = `
<div>
  <div
    class="css-1q24bon-Header"
  >
    <div
      class="css-heqg68-Header"
    >
      <div
        class="css-e890g1-Header"
      >
        <h1
          class="css-u111td-Header"
        >
          CoinTrack
        </h1>
      </div>
      <div>
        <div
          class="css-z5adpb-FiatDropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-0-label downshift-0-toggle-button"
            class="css-ax4dup"
            id="downshift-0-toggle-button"
            type="button"
          >
            USD
            <svg
              class="css-yi2kl5-FiatDropdown"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </button>
        </div>
        <ul
          aria-labelledby="downshift-0-label"
          class="css-nngvzr-FiatDropdown"
          id="downshift-0-menu"
          role="listbox"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</div>
`;
